### PR TITLE
feat: version-aware vLog GC for MVCC (Phase 10)

### PR DESCRIPTION
**Pull request by ben1009**
## Summary

Store full internal keys (user_key + ts) in vLog entries so GC can distinguish between versions of the same key. Previously vLog stored only user keys, making it impossible for GC to know which version a vLog entry belonged to.

## Changes

**vLog key format**: When `TS_ENABLED`, vLog entries now store the full encoded internal key (`[memcomparable_user_key][!ts: u64 BE]`) instead of just the user key. No backward compatibility — old entries treated as dead.

**Read path key threading**:
- `lookup_memtable` / `lookup_sst_raw` return `(value, kind, found_key)` 3-tuple
- Found internal key flows through to `vlog.read()` for verification
- `get()` / `get_with_ts()` pass found key to `resolve_value()`

**Version-specific GC**:
- `get_with_kind_at_ts(user_key, ts)` — exact internal key lookup (no version scan)
- `compare_and_set_batch_at_ts()` — version-aware CAS that swaps the correct version pointer
- `gc.rs: check_liveness()` — uses `get_with_kind_at_ts` when TS enabled
- `gc.rs: compact_file()` — uses version-aware CAS when TS enabled

**New APIs**:
- `mem_table::get_raw_exact()` — exact internal key lookup, skips bloom
- `mem_table::get_versioned_raw_with_key()` — returns (value, found_key)
- `lsm_storage::get_with_kind_at_ts()` — point lookup at specific ts
- `lsm_storage::compare_and_set_batch_at_ts()` — batch version-aware CAS
- `vlog::index::lookup_by_user_key()` — find all versions of a user key in index

## Verification

- [x] `./dev check` passes (fmt, clippy, machete, typos)
- [x] 394/394 tests pass
- [x] 4 new GC integration tests cover: old version dropping, version preservation for active readers, cross-version entries, internal key storage
- [x] 3-round subagent review completed (correctness, edge cases, style/perf)

## Summary by CodeRabbit

* **New Features**
  * Timestamp-aware reads and version-aware batch compare-and-set for GC rewrites.

* **Improvements**
  * Stronger vLog validation using exact stored internal keys across read paths and table files.
  * Consistent cross-tier version handling (memtable, immutable memtables, SSTs, vLog).
  * vLog GC preserves live versions while reclaiming reclaimable data when safe.
  * vLog index supports user-key based lookups and MVCC-aware entries.

* **Tests**
  * Added integration tests covering MVCC-aware vLog GC and multi-version behavior.

* **Documentation**
  * Updated roadmap/todo to reflect MVCC progress.

### CODE REVIEW

File: kv-engine/src/lsm_storage.rs [hunk 1]
```diff
@@ -1432,6 +1509,81 @@ impl LsmStorageInner {
...
+            }
+        }
+
+        if !writes.is_empty() {
+            let raw_refs: Vec<(KeySlice, &[u8])> = writes
+                .iter()
+                .map(|(k, v)| (KeySlice::from_slice(k), v.as_slice()))
+                .collect();
+            state.memtable.put_raw_batch(&raw_refs)?;
+        }
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/lsm_storage.rs:1602]: Fixed in commit 037f69d. Added `put_raw_batch_no_wal` to `MemTable` that bypasses WAL — only updates skiplist, bloom filter, and approximate size. `compare_and_set_batch_at_ts` now uses it for idempotent GC rewrites. Shared insert logic via `put_raw_batch_inner(data, write_wal)`.

File: kv-engine/src/mem_table.rs [hunk 2]
```diff
@@ -192,6 +192,16 @@ impl MemTable {
...
 
+    /// Exact lookup by full encoded internal key (user key + timestamp).
+    /// Used by GC to check if a specific version's pointer still matches.
+    /// Skips the bloom filter because the key includes a timestamp suffix
+    /// that would produce a different hash than the user key alone.
+    pub fn get_raw_exact(&self, encoded_internal_key: &[u8]) -> Option<Bytes> {
+        self.map
+            .get(encoded_internal_key)
+            .map(|x| x.value().clone())
+    }
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/mem_table.rs:216]: Fixed in commit 037f69d. Implemented `put_raw_batch_no_wal` via shared `put_raw_batch_inner(data, write_wal)` helper — gates WAL write on `write_wal` flag. Used by `compare_and_set_batch_at_ts` for idempotent GC rewrites.

File: kv-engine/src/lsm_storage.rs [hunk 3]
```diff
@@ -842,20 +850,71 @@ impl LsmStorageInner {
...
+            for id in sst_ids {
+                if let Some(s) = state.sstables.get(id)
+                    && let Some((raw, found_key)) =
+                        s.point_get_with_hash_and_key(&encoded, bloom_hash, None)?
+                    && found_key == encoded
+                {
+                    return Ok(Self::parse_value_kind(raw));
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/lsm_storage.rs:917]: Fixed in commit 02e3523. Replaced the full level scan with `partition_point` on `encoded_user_key_prefix` — now only checks the candidate SST (and verifies `last_key` range), matching the pattern used in `lookup_sst_raw`.

File: kv-engine/src/tests/vlog_integration_tests/gc.rs [hunk 4]
```diff
@@ -310,3 +311,137 @@ fn test_gc_analyze_uses_index() {
...
+    let results = gc.gc_all().unwrap();
+
+    // At least one GC result should indicate entries were reclaimed
+    // The old file is always scheduled for deletion; if all entries were dead,
+    // keys_rewritten=0 and the file is deleted. Either way, GC ran.
+    assert!(!results.is_empty(), "GC should have processed vLog files");
+    // The old vLog file (file 0) should have been scheduled for deletion
+    // After GC, reading at latest ts should still return the newest value
+    let val = storage.get(b"foo").unwrap();
+    assert_eq!(val, Some(Bytes::from(vec![b'B'; 64])));
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/tests/vlog_integration_tests/gc.rs:384]: Fixed in commit 02e3523. Removed the `assert!(!results.is_empty())` — the test now validates the observable end state (newest value reads back correctly) regardless of whether reclamation happened during compaction or GC.

File: kv-engine/src/vlog/gc.rs [hunk 5]
```diff
@@ -170,8 +170,27 @@ impl<'a> GarbageCollector<'a> {
...
+            // version-specific lookup.
+            match (
+                crate::key::decode_user_key(key),
+                crate::key::extract_ts(key),
+            ) {
+                (Some(user_key), Some(ts)) => self.inner.get_with_kind_at_ts(&user_key, ts)?,
+                _ => {
+                    // Fallback for malformed keys
+                    self.inner.get_with_kind(key)?
+                }
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/vlog/gc.rs]: Fixed in commit 02e3523. Malformed/legacy keys (decode_user_key or extract_ts fails) now `return Ok(false)` — treated as dead per the PR's no-backward-compat policy.

File: todo.md [hunk 6]
```diff
@@ -114,7 +114,7 @@ PR #82 (merged 2026-06-10). Optimistic concurrency control for serializable isol
 
 ---
 
-## Testing Progress (20/30 from RFC §9)
+## Testing Progress (23/30 from RFC §9)
```
[the remark this answers is not in the document]
Comment (ben1009) [reply todo.md]: Fixed in commit 02e3523. Updated count to 26/30.

**Comment by ben1009** [comment]
All review comments have been addressed and resolved. Please re-review.

@gemini-code-assist @coderabbitai

**Comment by ben1009** [comment]
Addressed Gemini review feedback (commit c1d07dd):

1. **mem_table.rs:203** — `get_raw_exact` now checks the bloom filter using the decoded user key before skiplist lookup. Skips expensive lookups when key is not in this memtable.

2. **vlog/index.rs:86** — `lookup_by_user_key` now compares raw `encoded_user_key_prefix` slices instead of decoding user keys per entry. Eliminates N heap allocations per lookup.

@gemini-code-assist please re-review.

**Comment by ben1009** [comment]
Fixed in commit ab8abe4. The leveled SST lookup now loops backwards from idx (matching lookup_sst_raw pattern) to handle user keys split across adjacent SSTs. Breaks when last_key prefix < target prefix.

@gemini-code-assist please re-review.

**Comment by ben1009** [comment]
Fixed in commit bfb712c. Replaced `unwrap_or_else`/`unwrap_or` with `.expect()` — keys from live entries are guaranteed well-formed by `check_liveness`.

@gemini-code-assist please re-review.
